### DISCUSSION TITLE
Get http_code=400 in rgw-rooks3 for PUT request from velero

### DISCUSSION THREAD

**Discussion by dajohb** [rook/rook#14475] [General]
Hi all,
hope someone can help me.
Since update to rook 1.14.8 i get this error for rook-s3.
My Velero backup try to upload a file with PUT requests but all failed with error htt_status=400.
here the log for example:

debug 2024-07-20T18:02:08.976+0000 7f6f5e69e700  1 ====== starting new request req=0x7f6eb21be730 =====
debug 2024-07-20T18:02:08.977+0000 7f6f5e69e700  1 ====== req done req=0x7f6eb21be730 op status=-2040 http_status=400 latency=0.001000002s ======
debug 2024-07-20T18:02:08.977+0000 7f6f5e69e700  1 beast: 0x7f6eb21be730: 10.244.11.1 - backup2s3 [20/Jul/2024:18:02:08.976 +0000] "PUT /rooks3/backups/backup-namespace-default-testing/velero-backup.json?x-id=PutObject HTTP/1.1" 400 3053 - "aws-sdk-go-v2/1.24.1 os/linux lang/go#1.22.2 md/GOOS#linux md/GOARCH#amd64 api/s3#1.48.0 ft/s3-transfer" - latency=0.001000002s
debug 2024-07-20T18:02:13.406+0000 7f6f8d6fc700  1 ====== starting new request req=0x7f6eb213d730 =====

any ideas?

Thank you very much

**Comment by BlaineEXE** [comment]
Has this worked in the past and is now failing, or is this a new attempt? 

In the PUT log I don't see a server endpoint, which could be an issue. But I'm not certain.

Ceph slack users might have more experience with these logs and how to more quickly debug error 400: https://ceph.io/en/community/connect/

**Reply by dajohb** [reply]
Hi,
thank you very much for the response.
What do you mean with the endpoint.
The log is from the k8 deployment rook-ceph-rgw-rooks3-a

**Reply by BlaineEXE** [reply]
@thotz might have some experience reading these Ceph RGW logs to suggest what may be the source of the 400 error.

**Reply by thotz** [reply]
2040 points to ERR_AMZ_CONTENT_SHA256_MISMATCH in RGW, mostly aws s3 v4 signature mismatch related errors. can you use the latest ceph images of squid or reef and see whether it resolves? A lot of fixes related to aws s3 v4 signature mismatch went in AFAIR

**Reply by dajohb** [reply]
i already use the latest version of rook.

by the way, im wondering why this PUT request to s3 works:

debug 2024-07-24T13:59:50.412+0000 7f6ed3588700  1 beast: 0x7f6eb2341730: 10.244.15.0 - backup2s3 [24/Jul/2024:13:59:50.404 +0000] "PUT /rooks3/restic/default/locks/213769297a6973cc1232f00dbd3baec8826abffeeaabeab479d034fd0e9fb5a2 HTTP/1.1" 200 149 - "MinIO (linux; amd64) minio-go/v7.0.46" - latency=0.008000016s

But this failed
debug 2024-07-24T13:59:50.662+0000 7f6f60ea3700  1 beast: 0x7f6eb22c0730: 10.244.11.1 - backup2s3 [24/Jul/2024:13:59:50.662 +0000] "PUT /rooks3/backups/backup-namespace-default-testing2/backup-namespace-default-testing2-logs.gz?x-id=PutObject HTTP/1.1" 400 10321 - "aws-sdk-go-v2/1.24.1 os/linux lang/go#1.22.2 md/GOOS#linux md/GOARCH#amd64 api/s3#1.48.0 ft/s3-transfer" - latency=0.000000000s

**Reply by BlaineEXE** [reply]
@dajohb, Jiffin is referring to the Ceph version. Which Ceph version is being used in your cluster?

**Reply by dajohb** [reply]
Hi all,

The ceph version is

 kubectl -n rook-ceph get deployment -l rook_cluster=rook-ceph -o jsonpath='{range .items[*]}{"ceph-version="}{.metadata.labels.ceph-version}{"\n"}{end}'

ceph-version=18.2.2-0

i've updated to latest version of rook and ceph based on this documentation:

https://rook.io/docs/rook/latest-release/Upgrade/ceph-upgrade/#2-update-the-toolbox-image

**Reply by 13CSherman** [reply]
> 2040 points to ERR_AMZ_CONTENT_SHA256_MISMATCH in RGW, mostly aws s3 v4 signature mismatch related errors. can you use the latest ceph images of squid or reef and see whether it resolves? A lot of fixes related to aws s3 v4 signature mismatch went in AFAIR

Just wondering if you have a link to a page that shows what these op status codes mean? I have a op status 2021 I'm trying to chase down.

**Reply by BlaineEXE** [reply]
This and below define the RGW codes afaict: https://github.com/ceph/ceph/blob/7ab070413cff8a6954142fa7855d6d258822eadf/src/rgw/rgw_common.h#L274

**Comment by thotz** [comment]
@dajohb please increase the rgw level log to 20 via following,
From toolbox pod fetch ceph rgw authx client name,` ceph auth ls | grep <for objectstore name>`
Then set the following `ceph config set <rgw authx name> debug_rgw 20`. This gives what is difference between valero put request with minio client put request. 

Collect those logs and  file an issue in ceph tracker with https://tracker.ceph.com/

**Reply by dajohb** [reply]
thank you,
here the log
`debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s s3:put_obj rgw::auth::s3::AWSAuthStrategy granted access
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj normalizing buckets and tenants
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s s->object=backups/backup-namespace-default-testing3/velero-backup.json s->bucket=rooks3
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj init permissions
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s s3:put_obj cache get: touching lru, lru_counter=395301 promotion_ts=395296
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s s3:put_obj moving rooks3.rgw.meta+root+rooks3 to cache LRU end
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s s3:put_obj cache get: name=rooks3.rgw.meta+root+rooks3 : hit (requested=0x11, cached=0x11)
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 15 req 13662189528033870194 0.001000004s s3:put_obj decode_policy Read AccessControlPolicy<AccessControlPolicy xmlns="http://s3.amazonaw
s.com/doc/2006-03-01/"><Owner><ID>backup2s3</ID><DisplayName>Backup User</DisplayName></Owner><AccessControlList><Grant><Grantee xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" x
si:type="CanonicalUser"><ID>backup2s3</ID><DisplayName>Backup User</DisplayName></Grantee><Permission>FULL_CONTROL</Permission></Grant></AccessControlList></AccessControlPolicy>
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s s3:put_obj cache get: touching lru, lru_counter=395302 promotion_ts=395297
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s s3:put_obj moving rooks3.rgw.meta+users.uid+backup2s3 to cache LRU end
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s s3:put_obj cache get: name=rooks3.rgw.meta+users.uid+backup2s3 : hit (requested=0x13, cached=0x
13)
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj recalculating target
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj reading permissions
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj init op
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj verifying op mask
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s s3:put_obj required_mask= 2 user.op_mask=7
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj verifying op permissions
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s get_encryption_defaults:  kms_attr_seen is 0 and sse_algorithm is 
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s s3:put_obj -- Getting permissions begin with perm_mask=50
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj Searching permissions for identity=rgw::auth::SysReqApplier -> rgw::auth::LocalApplier(acct_user=backup2s3, acct_name=Backup User, subuser=, perm_mask=15, is_admin=0) mask=50
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj Searching permissions for uid=backup2s3
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj Found permission: 15
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj Searching permissions for group=1 mask=50
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj Permissions for group not found
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj Searching permissions for group=2 mask=50
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj Permissions for group not found
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj -- Getting permissions done for identity=rgw::auth::SysReqApplier -> rgw::auth::LocalApplier(acct_user=backup2s3, acct_name=Backup User, subuser=, perm_mask=15, is_admin=0), owner=backup2s3, perm=2
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s s3:put_obj  identity=rgw::auth::SysReqApplier -> rgw::auth::LocalApplier(acct_user=backup2s3, acct_name=Backup User, subuser=, perm_mask=15, is_admin=0) requested perm (type)=2, policy perm=2, user_perm_mask=2, acl perm=2
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj verifying op params
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj pre-executing
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj check rate limiting
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj executing
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s cache get: touching lru, lru_counter=395303 promotion_ts=395298
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s moving rooks3.rgw.log++pubsub..bucket.rooks3/9a2ddd81-5628-4fe4-8366-06f8decc9f97.63411.2 to cache LRU end
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s cache get: name=rooks3.rgw.log++pubsub..bucket.rooks3/9a2ddd81-5628-4fe4-8366-06f8decc9f97.63411.2 : hit (negative entry)
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s s3:put_obj max_chunk_size=4194304
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s cache get: touching lru, lru_counter=395304 promotion_ts=395299
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s moving rooks3.rgw.log++script.putdata. to cache LRU end
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s cache get: name=rooks3.rgw.log++script.putdata. : hit (negative entry)
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  5 req 13662189528033870194 0.001000004s s3:put_obj NOTICE: call to do_aws4_auth_completion
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR: x-amz-content-sha256 does not match
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR:   grab_aws4_sha256_hash()=022e8f9abbdd7f8f08a076c4c8159dc8a42de571b39d1d810f3822c721f32196
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR:   expected_request_payload_hash=STREAMING-UNSIGNED-PAYLOAD-TRAILER
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s s3:put_obj get_data() returned ret=-2040
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj completing
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s cache get: touching lru, lru_counter=395305 promotion_ts=395300
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s moving rooks3.rgw.log++script.postrequest. to cache LRU end
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 req 13662189528033870194 0.001000004s cache get: name=rooks3.rgw.log++script.postrequest. : hit (negative entry)
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj op status=-2040
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  2 req 13662189528033870194 0.001000004s s3:put_obj http status=400
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  1 ====== req done req=0x7f6eb223f730 op status=-2040 http_status=400 latency=0.001000004s ======
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700  1 beast: 0x7f6eb223f730: 10.244.15.0 - backup2s3 [25/Jul/2024:06:42:37.536 +0000] "PUT /rooks3/backups/backup-namespace-default-testing3/velero-backup.json?x-id=PutObject HTTP/1.1" 400 3054 - "aws-sdk-go-v2/1.24.1 os/linux lang/go#1.22.2 md/GOOS#linux md/GOARCH#amd64 api/s3#1.48.0 ft/s3-transfer" - latency=0.001000004s
debug 2024-07-25T06:42:40.744+0000 7f6fb8752700 20 lua background: cache get: touching lru, lru_counter=395306 promotion_ts=395221
debug 2024-07-25T06:42:40.744+0000 7f6fb8752700 10 lua background: moving rooks3.rgw.log++script.background. to cache LRU end
debug 2024-07-25T06:42:40.744+0000 7f6fb8752700 10 lua background: cache get: name=rooks3.rgw.log++script.background. : hit (negative entry

`

how you i set the "normal" Loglevel again? Or what is the default Loglevel? :)

**Reply by thotz** [reply]
set to 1 for the normal config. The following are the errors.
```
do_aws4_auth_completion
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR: x-amz-content-sha256 does not match
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR: grab_aws4_sha256_hash()=022e8f9abbdd7f8f08a076c4c8159dc8a42de571b39d1d810f3822c721f32196
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR: expected_request_payload_hash=STREAMING-UNSIGNED-PAYLOAD-TRAILER
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 20 req 13662189528033870194 0.001000004s s3:put_obj get_data() returned ret=-2040
```
 Either it may be issue RGW or with valero client, Please file issue in https://tracker.ceph.com/

**Reply by mattbenjamin** [reply]
this is a streaming-unsigned payload case, but this log pattern I think indicates we are in AWSv4ComplSingle::complete().  I don't think we handle payload trailer in the single case;  it would be helpful to have a concise reproducer for this

**Reply by mattbenjamin** [reply]
ok, the failure is in an aws sdk client request, but it worked with the minio go client;  there's always potential variation

**Reply by mattbenjamin** [reply]
it might be a behavior introduced in a recent update to the awssdk-go-v2 client.  specifically, it seems like the scenario is the the client is sending a trailing header/header section but is not using the aws-chunked transfer encoding.   It seems like experimentation with the latest sdk might turn it up

**Reply by cbodley** [reply]
@mattbenjamin from the log output, it looks like this doesn't have your STREAMING-UNSIGNED-PAYLOAD-TRAILER changes:
```
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR: x-amz-content-sha256 does not match
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR: grab_aws4_sha256_hash()=022e8f9abbdd7f8f08a076c4c8159dc8a42de571b39d1d810f3822c721f32196
debug 2024-07-25T06:42:37.537+0000 7f6f33e49700 10 ERROR: expected_request_payload_hash=STREAMING-UNSIGNED-PAYLOAD-TRAILER
```
this hadn't been backported for `ceph-version=18.2.2-0`

**Reply by cbodley** [reply]
the reef backport https://tracker.ceph.com/issues/64465 still hasn't merged. it should show up in 18.2.5

**Comment by mattbenjamin** [comment]
a larger chunk of the log with debug info during the failure might help see what induces the sdk to send this format
